Get-DbaConfiguration - reads sys.configuration data using SMO

sirsql opened the pull request:
Fixes # 

Changes proposed in this pull request:
 - New function that uses SMO to read the sp_configure/sys.configurations data


How to test this code: 
- [ ] Get-DbaConfiguration -SqlServer localhost
- [ ] 

Has been tested on minimum requirements:
- [ ]  Powershell 3
- [ ]  Windows 7
- [ ]  SQL Server 2000

Has been tested on maximum requirements:
- [ ]  SQL Server vNext
- [ ]  Windows 10
- [ ]  Azure Database

Tests for tester:
- [ ] Working/useful help content, including link to command on dbatools web site
- [ ] All examples work as advertised
- [ ] Does not contain template content
- [ ] Does not contain excessive/unnecessary amounts of comments
- [ ] Works remotely
- [ ] Works locally
- [ ] Works on lower versions or throws error specifying version not supported
- [ ] Works with named instances
- [ ] Works with clustered instances
- [ ] Handles offline/read only databases
- [ ] Supports multiple servers (at the command line or piped from Get-SqlRegisteredServerName)
- [ ] No un-handled errors which stop the command working with multiple servers

PowerDBAKlaas commented on functions/Get-DbaConfiguration.ps1:
```diff
@@ -0,0 +1,108 @@
+﻿FUNCTION Get-DbaConfiguration
```
I feel we need to put "sql" or "instance" somewhere in the name, because the OS and other stuff have configurations too.

sirsql replied:
Get-DbaInstanceConfiguration be better?
Would indicate that these are instance level configs, rather than the database level configs that exist in 2016+

ctrlbold replied:
I based the name off of the proposed name from the official Microsoft board, which was likely based off of SMO. I'm good with keeping it. Alternatively, we can do what I did in a few other commands: Get-DbaSpConfigure 

Copy-SqlSpConfigure
Export-SqlSpConfigure
Import-SqlSpConfigure

Not a huge fan of Instance but have no discernible reason.

sirsql replied:
Updated function to Get-DbaSpConfigure

PowerDBAKlaas commented on functions/Get-DbaConfiguration.ps1:
```diff
@@ -0,0 +1,108 @@
+				}
+			}
+			
+			if ($Detailed -eq $true)
```
I feel when we have 'only' 10 properties, we shouldn't have a -detailed switch.
 This goes for some existing functions too.

sirsql replied:
Happy to remove.

sirsql replied:
Removed switch

PowerDBAKlaas commented on functions/Get-DbaConfiguration.ps1:
```diff
@@ -0,0 +1,108 @@
+			
+			if ($Detailed -eq $true)
+			{
+				$output | Sort-Object ServerName, DisplayName
```
We must avoid sorting inside a function.
It comes with a performance penalty, 
maybe the user doesn't care about the order, or
worse: he'll pipe again to sort-object

sirsql replied:
Will remove the sort.

sirsql replied:
Sort removed in latest commit

PowerDBAKlaas commented on functions/Get-DbaConfiguration.ps1:
```diff
@@ -0,0 +1,108 @@
+			$objList = get-member -InputObject $server.Configuration -MemberType Property -Force | select Name | where { $_.Name -ne "Parent" -and $_.Name -ne "Properties" }
+			
+			#Iterate through the properties to get the configuration settings
+			foreach ($prop in $objList)
```
I don't get the reason for the looping, the if, nor the propinfo.

wouldn't
```
$server.configuration.Properties |
 SELECT @{N="ComputerName";e={$server.NetName}},DisplayName, RunValue,...
```
suffice?

Unless I'm missing something, the next 30 lines are not needed.

sirsql replied:
The looping was needed to handle returning the parent property name (which is how the configuration value is stored in SMO). There is no parent property object in $Server.configuration.properties, so this was the only way to handle that.

PowerDBAKlaas commented on functions/Get-DbaConfiguration.ps1:
```diff
@@ -0,0 +1,108 @@
+	)
+	BEGIN
+	{
+		$output = @()
```
We shouldn't collect objects, but put them immediately in the pipeline.

sirsql replied:
Are there any examples I could take a look at for this?

ctrlbold replied:
This new PR does it -- #508

sirsql replied:
Changes to push directly to the pipeline

ctrlbold commented on functions/Get-DbaConfiguration.ps1:
```diff
@@ -0,0 +1,108 @@
+			CATCH
+			{
+				Write-Verbose "Failed to connect to: $Server"
+				continue
```
this should be write-warning

sirsql replied:
Fixed

ctrlbold commented on functions/Get-DbaConfiguration.ps1:
```diff
@@ -0,0 +1,108 @@
+			}
+			
+			#Get a list of the configuration property parents, and exlude the Parent, Properties values
+			$objList = get-member -InputObject $server.Configuration -MemberType Property -Force | select Name | where { $_.Name -ne "Parent" -and $_.Name -ne "Properties" }
```
I would actually just do  $server.Configuration 

When I make Get-DbaDatabase, I'll send $server.Databases immediately to the pipeline -- it looks like that's what Get-SqlDatabase does too, actually. So ultimately, this command can have as few as 5 lines. I'll submit a PR for Get-DbaDatabase so you can see. 

@PowerDBAKlaas please review my next command. I'd like to make sure we're on the same page.

@sirsql thanks so much and welcome to the team! 👍 💯 🥇

ctrlbold commented on functions/Get-DbaConfiguration.ps1:
```diff
@@ -0,0 +1,108 @@
+{
+<#
+.SYNOPSIS
+Returns all server level system configuration (sys.configuration/sp_configure) information
```
I really like your docs btw!

ctrlbold commented on functions/Get-DbaConfiguration.ps1:
```diff
@@ -0,0 +1,108 @@
+	}
+	PROCESS
+	{
+		FOREACH ($Server in $SqlServer)
```
I also noticed-- this should be named differently. like foreach ($instance in $sqlserver) 


then a few lines below, $server = connect-sqlserver -sqlserver $instance

sirsql replied:
Fixed

sirsql commented:
Several changes based on feedback.

- Changed name to Get-DbaSpConfigure to better reflect what it pulls
- Outputs directly to the pipeline, no more loading up a variable
- Removed the sorts
- Removed the detailed switch, now returns all columns
- Fixed the write-verbose so that it is now a write-warning
- Fixed case where $server was being used incorrectly

Have left in the loop as it's the only way to capture the parent of the configuration, and this is important should a user wish to set them.Important reference point.

PowerDBAKlaas commented:
Ha, now I understand what you're saying.
It's the configname.
Of course that's a valid reason.

@ctrlbold we leave the ComputerName - SqlInstance -InstanceName for march, right?

ctrlbold commented:
Naw, I just accepted it ;) March is right around the corner and I didn't have time to switch everything back. I'll update the website banner when i get a chance.

PowerDBAKlaas commented:
what do you mean? It should be changed now?

ctrlbold commented:
The code is good as is. And we'll accept PRs with both SqlServer and SqlInstance then standardize in march.

PowerDBAKlaas commented:
Hi Nic

I have an additional question: wouldn't it be nice to have an extra property 'IsDefault'?
$false for every configuration.property where [RunningValue] or [ConfiguredValue] is not the DefaultValue.
That may be a handy property to filter on for a review,
and I see that's what sp_Blitz and SSMS standard reports do too.

sirsql commented:
That's a great idea, I'll add that.

wsmelton commented:
@sirsql if you want I had written code that contained all the default values of SQL Server versions (pulled from BOL) a few years ago. We could build this out into an internal function to store that info, then utilize in this command to more easily find the "default".

[blog article](http://blog.wsmelton.info/2013-12-04-ps-check-instanceconfig-part-1/)
[Actual code](https://github.com/wsmelton/scripts/blob/master/posh/Check-InstanceConfig.ps1)

Mind you it was years ago so it probably can be cleaned up and optimized.

sirsql commented:
Thank @wsmelton I was just checking and finding out that the server default values are not stored, so a function like that will be really useful. I'll have a dig through your code.

sirsql commented:
Added the default configurations for each version in a new internal function. Added a call to that in the Get-DbaSpConfigure function, now returns the defaults, and whether or not the running value matches the default.

ctrlbold commented:
Fantastic, useful command with wonderful output! Great teamwork too. Ran it against all my servers from 2000 to vnext with success! Wonderful work, @sirsql
